Detectors/TRD/reconstruction/src/EventRecord.cxx

```diff
@@ -44,192 +40,58 @@
-{
-  bool added = false;
-  int count = 0;
-  for (int count = 0; count < mEventRecords.size(); ++count) {
-    if (ir == mEventRecords[count].getBCData()) {
-      //TODO replace this with a hash/map not a vector
-      mEventRecords[count].addTracklets(tracklets); //mTracklets.insert(mTracklets.back(),start,end);
-      added = true;
-    }
-  }
-  if (!added) {
-    // unseen ir so add it
-    mEventRecords.push_back(ir);
-    mEventRecords.back().addTracklets(tracklets);
-  }
-}
-void EventStorage::addTracklets(InteractionRecord& ir, std::vector<Tracklet64>::iterator& start, std::vector<Tracklet64>::iterator& end)
-{
-  bool added = false;
-  for (int count = 0; count < mEventRecords.size(); ++count) {
-    if (ir == mEventRecords[count].getBCData()) {
-      //TODO replace this with a hash/map not a vector
-      mEventRecords[count].addTracklets(start, end); //mTracklets.insert(mTracklets.back(),start,end);
-      added = true;
-    }
-  }
-  if (!added) {
-    // unseen ir so add it
-    mEventRecords.push_back(ir);
-    mEventRecords.back().addTracklets(start, end);
-    //  LOG(info) << "x unknown ir adding " << std::distance(start,end)<< " tracklets";
-  }
-}
-void EventStorage::unpackData(std::vector<TriggerRecord>& triggers, std::vector<Tracklet64>& tracklets, std::vector<Digit>& digits)
-{
-  int digitcount = 0;
-  int trackletcount = 0;
-  for (auto& event : mEventRecords) {
-    tracklets.insert(std::end(tracklets), std::begin(event.getTracklets()), std::end(event.getTracklets()));
-    digits.insert(std::end(digits), std::begin(event.getDigits()), std::end(event.getDigits()));
-    triggers.emplace_back(event.getBCData(), digitcount, event.getDigits().size(), trackletcount, event.getTracklets().size());
-    digitcount += event.getDigits().size();
-    trackletcount += event.getTracklets().size();
-  }
-}
-
-void EventStorage::sendData(o2::framework::ProcessingContext& pc, bool generatestats)
+void EventRecordContainer::sendData(o2::framework::ProcessingContext& pc, bool generatestats)
 {
   //at this point we know the total number of tracklets and digits and triggers.
   auto dataReadStart = std::chrono::high_resolution_clock::now();
-  uint64_t trackletcount = 0;
-  uint64_t digitcount = 0;
-  uint64_t triggercount = 0;
-  sumTrackletsDigitsTriggers(trackletcount, digitcount, triggercount);
+
+  size_t digitcount = 0;
+  size_t trackletcount = 0;
   std::vector<Tracklet64> tracklets;
-  tracklets.reserve(trackletcount);
   std::vector<Digit> digits;
-  digits.reserve(digitcount);
   std::vector<TriggerRecord> triggers;
-  triggers.reserve(triggercount);
   for (auto& event : mEventRecords) {
-    //sort tracklets
-    event.sortByHCID();
-    //TODO do this sort in parallel over the events
-    tracklets.insert(std::end(tracklets), std::begin(event.getTracklets()), std::end(event.getTracklets()));
-    digits.insert(std::end(digits), std::begin(event.getDigits()), std::end(event.getDigits()));
+    event.sortTrackletsByHCID();
```

Comment:
Not sure if this could be done so easily. The data comes in HBFs from single half CRUs. So in one chunk we always have multiple triggers for different links. If we would want to keep only single vectors of digits and tracklets we would need to do multiple checks when inserting (for trigger number and for detector ID). So this might turn out more processing intensive in the end. But that's just an assumption, I have not tested it.

Reply:
data is read is in as : 
1a 1b 1c 1d 1e 1f 2a 2c 2d 2e 2f 3b 3c 3d 3e 3f ..... 1079 1079c 1079f 1g 1h 1i 1j 2g 2h 2i 2j .... 
a being event and integer being hcid, the hcid order is not sequential, but for the purposes of this it will suffice.

You dont know how big each block is, so either you bin them by 1,2,3,4 or you bin them by a,b,c,d,e,f
leading to a store of each event, which when finished is then merged into the message memory space via the copies above.
Now one could change the trigger record  format to include a range of ranges, but that would mess with a lot of code.
Or one could have an additional dpl device to merge that all together, but why.
Now there are other copies that one can remove but I doubt you will get past this one.